File: p2p_library/src/main/java/com/tambapps/p2p/file_sharing/FileSharer.java
package com.tambapps.p2p.file_sharing;

import com.tambapps.p2p.file_sharing.concurrent.FutureShare;
import com.tambapps.p2p.file_sharing.concurrent.ShareCallable;
import com.tambapps.p2p.file_sharing.task.FileProvider;
import com.tambapps.p2p.file_sharing.task.ReceivingTask;
import com.tambapps.p2p.file_sharing.task.SendingTask;
import com.tambapps.p2p.file_sharing.util.FileUtils;

import java.io.File;
import java.io.IOException;
import java.util.concurrent.*;

/**
 * Util class to share files without having to deals with tasks
 */
public class FileSharer {

  private final ExecutorService executorService;

  public FileSharer() {
    this(Runtime.getRuntime().availableProcessors());
  }

  public FileSharer(int nbThreads) {
    this.executorService = Executors.newFixedThreadPool(nbThreads);
  }

  public Future<Boolean> sendFile(String filePath, Peer peer) {
    return sendFile(new File(filePath), peer, SendingTask.DEFAULT_SOCKET_TIMEOUT);
  }

  public Future<Boolean> sendFile(File file, Peer peer) {
    return sendFile(file, peer, SendingTask.DEFAULT_SOCKET_TIMEOUT);
  }

  public Future<Boolean> sendFile(File file, Peer peer, int socketTimout) {
    if (!file.isFile()) {
      throw new IllegalArgumentException(file.getPath() + " isn't a file");
    }
    SendCallable callable = new SendCallable(file, peer, socketTimout);
    return new FutureShare(executorService.submit(callable), callable);
  }

  public Future<Boolean> receiveFile(FileProvider fileProvider, Peer peer) {
    ReceiveCallable callable = new ReceiveCallable(fileProvider, peer);
    return new FutureShare(executorService.submit(callable), callable);
  }

  public Future<Boolean> receiveFile(File file, Peer peer) {
    return receiveFile(name -> file, peer);
  }

  public Future<Boolean> receiveFile(String filePath, Peer peer) {
    return receiveFile(new File(filePath), peer);
  }

  public Future<Boolean> receiveFileInDirectory(File directory, Peer peer) {
    if (!directory.exists()) {
      throw new IllegalArgumentException(directory + " doesn't exist");
    }
    if (!directory.isDirectory()) {
      throw new IllegalArgumentException(directory + " isn't a directory");
    }
    return receiveFile(name -> FileUtils.newAvailableFile(directory, name), peer);
  }

  static class SendCallable implements ShareCallable {
    private final SendingTask task;
    private File file;

    public SendCallable(File file, Peer peer, int socketTimout) {
      this.task = new SendingTask(peer, socketTimout);
      this.file = file;
    }

    @Override
    public void cancel() {
      task.cancel();
    }

    @Override
    public Boolean call() {
      try {
        task.send(file);
        return true;
      } catch (IOException e) {
        return false;
      }
    }
  }

  static class ReceiveCallable implements ShareCallable {

    private final ReceivingTask task;
    private Peer peer;

    public ReceiveCallable(FileProvider fileProvider, Peer peer) {
      this.task = new ReceivingTask(fileProvider);
      this.peer = peer;
    }

    @Override
    public void cancel() {
      task.cancel();
    }

    @Override
    public Boolean call() {
      try {
        task.receiveFrom(peer);
        return true;
      } catch (IOException e) {
        return false;
      }
    }
  }
}

File: p2p_library/src/main/kotlin/com/tambapps/p2p/fandem/FileSharer.kt
package com.tambapps.p2p.fandem

import com.tambapps.p2p.fandem.concurrent.FutureShare
import com.tambapps.p2p.fandem.concurrent.SharingCallable
import com.tambapps.p2p.fandem.listener.SharingErrorListener
import com.tambapps.p2p.fandem.listener.TransferListener
import com.tambapps.p2p.fandem.task.FileProvider
import com.tambapps.p2p.fandem.task.ReceivingTask
import com.tambapps.p2p.fandem.task.SendingTask
import com.tambapps.p2p.fandem.util.FileUtils
import java.io.File
import java.io.IOException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Future

/**
 * Util class to share files without having to deal with tasks
 */
class FileSharer(private val executorService: ExecutorService) {
  fun sendFile(filePath: String, peer: Peer, errorListener: SharingErrorListener): Future<Boolean> {
    return sendFile(File(filePath), peer, SendingTask.DEFAULT_SOCKET_TIMEOUT, null, errorListener)
  }

  fun sendFile(filePath: String, peer: Peer, transferListener: TransferListener?, errorListener: SharingErrorListener): Future<Boolean> {
    return sendFile(File(filePath), peer, SendingTask.DEFAULT_SOCKET_TIMEOUT, transferListener, errorListener)
  }

  fun sendFile(file: File, peer: Peer, errorListener: SharingErrorListener): Future<Boolean> {
    return sendFile(file, peer, SendingTask.DEFAULT_SOCKET_TIMEOUT, null, errorListener)
  }

  fun sendFile(file: File, peer: Peer, transferListener: TransferListener?, errorListener: SharingErrorListener): Future<Boolean> {
    return sendFile(file, peer, SendingTask.DEFAULT_SOCKET_TIMEOUT, transferListener, errorListener)
  }

  fun sendFile(file: File, peer: Peer, socketTimout: Int, transferListener: TransferListener?, errorListener: SharingErrorListener): Future<Boolean> {
    require(file.isFile) { file.path + " isn't a file" }
    val callable = SendCallable(file, peer, socketTimout, transferListener, errorListener)
    return FutureShare(executorService.submit(callable), callable)
  }

  fun receiveFile(fileProvider: FileProvider, peer: Peer, transferListener: TransferListener?, errorListener: SharingErrorListener): Future<Boolean> {
    val callable = ReceiveCallable(fileProvider, peer, transferListener, errorListener)
    return FutureShare(executorService.submit(callable), callable)
  }

  fun receiveFile(file: File, peer: Peer, errorListener: SharingErrorListener): Future<Boolean> {
    return receiveFile(FileUtils.staticFileProvider(file), peer, null, errorListener)
  }

  fun receiveFile(file: File, peer: Peer, transferListener: TransferListener, errorListener: SharingErrorListener): Future<Boolean> {
    return receiveFile(FileUtils.staticFileProvider(file), peer, transferListener, errorListener)
  }

  fun receiveFile(filePath: String, peer: Peer, errorListener: SharingErrorListener): Future<Boolean> {
    return receiveFile(File(filePath), peer, errorListener)
  }

  fun receiveFile(filePath: String, peer: Peer, transferListener: TransferListener, errorListener: SharingErrorListener): Future<Boolean> {
    return receiveFile(File(filePath), peer, transferListener, errorListener)
  }

  fun receiveFileInDirectory(directory: File, peer: Peer, transferListener: TransferListener?,
                             errorListener: SharingErrorListener): Future<Boolean> {
    require(directory.exists()) { "$directory doesn't exist" }
    require(directory.isDirectory) { "$directory isn't a directory" }
    return receiveFile(FileUtils.availableFileInDirectoryProvider(directory),
            peer, transferListener, errorListener)
  }

  fun receiveFileInDirectory(directory: File, peer: Peer, errorListener: SharingErrorListener): Future<Boolean> {
    return receiveFileInDirectory(directory, peer, null, errorListener)
  }
}

private class SendCallable internal constructor(private val file: File, peer: Peer, socketTimout: Int, transferListener: TransferListener?, private val errorListener: SharingErrorListener) : SharingCallable {
  private val task: SendingTask = SendingTask(transferListener, peer, socketTimout)
  override fun cancel() {
    task.cancel()
  }

  override fun call(): Boolean {
    return try {
      task.send(file)
      true
    } catch (e: IOException) {
      errorListener.onError(e)
      false
    }
  }
}

private class ReceiveCallable internal constructor(fileProvider: FileProvider, private val peer: Peer, transferListener: TransferListener?, private val errorListener: SharingErrorListener) : SharingCallable {
  private val task: ReceivingTask = ReceivingTask(transferListener, fileProvider)
  override fun cancel() {
    task.cancel()
  }

  override fun call(): Boolean {
    return try {
      task.receiveFrom(peer)
      true
    } catch (e: IOException) {
      errorListener.onError(e)
      false
    }
  }
}